### DISCUSSION TITLE
Using openssl 3.3 legacy provider on iOS

### DISCUSSION THREAD

**Discussion by Barnie3** [openssl/openssl#25793] [Q&A] [answered]
Hello everyone,

I'm working on a cross platform mobile software using kotlin multiplatform tools. It does some crypto operations, for iOS implementation it uses OpenSSL C api and implementations provided by libcrypto.3.dylib. The library is linked with few other components to an xcframework (using kotlin/native tool called Cinterop) and so far worked fine.

Recently I had to add an implementation of the Retail-MAC (ISO 9797 alg 3) to the framework. OpenSSL doesn't seem to have it, but it has DES, which the MAC algorithm is based on, so I made my own using OpenSSL's DES-ECB cipher.

Using plain DES on openssl 3+ requires loading legacy provider. This is done by calling OSSL_PROVIDER_load(null, "legacy") and then EVP_CIPHER_fetch(null, "DES-ECB", "provider=legacy") to get the algorithm implementation. The legacy provider does not load, however (and the evp_cipher_fetch returns null); It might be because the legacy provider looks to be a separate entity and not a part of libcrypto.3.dylib. The openssl build script creates a file called legacy.dylib, but I can't link it to the xcframework the same way as libcrypto - I get a linker error saying "can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB)".

Is there a way to link the legacy provider 'bundle' to an iOS xcframework and have it usable by libcrypto?

Kind Regards

**Comment by mattcaswell** [comment, marked as answer]
I'm not familiar with iOS to be able to advise on the correct way to do this. However, a workaround might be to build OpenSSL with the "no-module" config option. Then the legacy provider becomes a "built-in" provider (i.e. it is part of libcrypto). You still have to load it via `OSSL_PROVIDER_load` but it does not need the separate legacy.dylib file.

**Reply by Barnie3** [reply]
Thank you for the reply.

I set the config of the OpenSSL build to ios64-xcrun no-module and it indeed does not make the legacy.dylib, but loading of the legacy provider still fails in the same fashion as earlier.

When I set the config to use no-module and no-shared options (fully static build), then it works. However, I would like to know, if there is an option that would preserve libcrypto as dynamic library?

**Reply by t8m** [reply]
Using just `no-module` should work. Which version of OpenSSL are you using?

**Reply by Barnie3** [reply]
It's version 3.3.2

**Reply by t8m** [reply]
Are you sure you used the right libcrypto shared library when you tested the `no-module` build? Because `no-module` should work and we test such configuration (although not on iOS). Better do `make clean` before building with a new configuration from the same directory.

**Reply by pastey** [reply]
Maybe it also needs `enable-weak-ssl-ciphers` configuration option?
